src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/UI/InfoBars/VsInfoBarService.cs

```diff
@@ -0,0 +1,207 @@
+﻿// Licensed to the .NET Foundation under one or more agreements. The .NET Foundation licenses this file to you under the MIT license. See the LICENSE.md file in the project root for more information.
+
+using Microsoft.VisualStudio.Imaging.Interop;
+using Microsoft.VisualStudio.Shell;
+using Microsoft.VisualStudio.Shell.Interop;
+using Task = System.Threading.Tasks.Task;
+
+namespace Microsoft.VisualStudio.ProjectSystem.VS.UI.InfoBars;
+
+/// <summary>
+/// Implementation of <see cref="IInfoBarService"/> that pushes messages to the info bar attached to Visual Studio's main window.
+/// </summary>
+[Export(typeof(IInfoBarService))]
+internal sealed class VsInfoBarService : IInfoBarService
+{
+    // This component exists per configured project. We track entries at the global level here, and keep track of the
+    // projects related to each info bar within the entry. This allows us to close info bars when all projects that
+    // requested them are unloaded.
+    private static readonly List<InfoBar> s_entries = [];
+
+    private static IVsInfoBarHost? s_host;
+
+    private readonly UnconfiguredProject _project;
+    private readonly IProjectThreadingService _threadingService;
+    private readonly IVsShellServices _vsShellServices;
+    private readonly IVsUIService<SVsShell, IVsShell> _vsShell;
+    private readonly IVsUIService<SVsInfoBarUIFactory, IVsInfoBarUIFactory> _vsInfoBarFactory;
+
+    [ImportingConstructor]
+    public VsInfoBarService(
+        UnconfiguredProject project,
+        IUnconfiguredProjectTasksService unconfiguredProjectTasksService,
+        IProjectThreadingService threadingService,
+        IVsShellServices vsShellServices,
+        IVsUIService<SVsShell, IVsShell> vsShell,
+        IVsUIService<SVsInfoBarUIFactory, IVsInfoBarUIFactory> vsInfoBarFactory)
+    {
+        _project = project;
+        _threadingService = threadingService;
+        _vsShellServices = vsShellServices;
+        _vsShell = vsShell;
+        _vsInfoBarFactory = vsInfoBarFactory;
+
+        // Ensure we clean up when the project unloads. If it has already unloaded when
+        // we register this callback, it is called immediately.
+        unconfiguredProjectTasksService.UnloadCancellationToken.Register(() =>
+        {
+            lock (s_entries)
+            {
+                foreach (InfoBar entry in s_entries)
+                {
+                    entry.OnProjectClosed(_project);
+                }
+            }
+        });
+    }
+
+    public async Task ShowInfoBarAsync(string message, ImageMoniker image, CancellationToken cancellationToken, params ImmutableArray<InfoBarAction> actions)
+    {
+        if (await _vsShellServices.IsCommandLineModeAsync(cancellationToken))
```

Comment:
What's command line mode in VS

Reply:
If you run `devenv /build MySolution.sln` then VS runs without any UI and builds the solution. We call this "command line mode" and there's no UI. We track this in several places to both:

- Improve performance (avoiding work when there's no UI)
- Prevent hangs (due to waiting for things that will never be available in this mode)